File: Frontend/src/services/announcements.js
/**
 * Announcements API service.
 * All calls go through the shared fetch-based `api` client, which attaches
 * the JWT token automatically and handles 401 session expiry.
 */
import { api } from './api';

// ---- Query-string builder (same pattern as leave.js) ----------------------
function buildQs(params) {
  if (!params) return '';
  const entries = Object.entries(params).filter(
    ([, v]) => v !== undefined && v !== null && v !== '' && v !== false
  );
  if (!entries.length) return '';
  return '?' + entries
    .map(([k, v]) => encodeURIComponent(k) + '=' + encodeURIComponent(v))
    .join('&');
}

// ---- API surface ----------------------------------------------------------
export const announcementsApi = {
  // ---- Feed (employee / manager / admin feed) ----------------------------
  /** List announcements. Pass mode='manage' for the admin management table. */
  list: (params) =>
    api.get('/announcements' + buildQs(params)),

  /** Latest N announcements for dashboard widgets (System Narrative etc.).
   *  Returns a bare array sorted by pinned-first then newest, role-filtered. */
  latest: (limit = 5) =>
    api.get('/announcements/latest' + buildQs({ limit })),

  /** Count unread announcements for the current user. */
  unreadCount: () =>
    api.get('/announcements/unread-count'),

  /** Fetch a single announcement by id. */
  one: (id) =>
    api.get(`/announcements/${id}`),

  // ---- Admin CRUD --------------------------------------------------------
  /** Create a new announcement (draft). */
  create: (payload) =>
    api.post('/announcements', payload),

  /** Update an existing announcement. */
  update: (id, payload) =>
    api.put(`/announcements/${id}`, payload),

  /** Delete an announcement. */
  remove: (id) =>
    api.delete(`/announcements/${id}`),

  // ---- Lifecycle transitions ---------------------------------------------
  /** Publish a draft announcement (admin/hr only). */
  publish: (id) =>
    api.post(`/announcements/${id}/publish`),

  /** Archive an announcement (admin/hr only). */
  archive: (id) =>
    api.post(`/announcements/${id}/archive`),

  // ---- Engagement tracking -----------------------------------------------
  /** Mark an announcement as read for the current user. */
  markRead: (id) =>
    api.post(`/announcements/${id}/read`),

  /** Acknowledge a mandatory announcement. */
  acknowledge: (id) =>
    api.post(`/announcements/${id}/acknowledge`),

  // ---- Analytics ---------------------------------------------------------
  /** Read/ack stats for a single announcement (admin/hr/manager). */
  analytics: (id) =>
    api.get(`/announcements/${id}/analytics`),
};

// ---- Static config (kept here so all pages share the same source of truth)

export const ANNOUNCEMENT_CATEGORIES = [
  'Company Updates',
  'Holidays',
  'Payroll',
  'Policies',
  'IT Maintenance',
  'Events',
  'Compliance',
  'Emergency Alerts',
  'Training',
];

export const ANNOUNCEMENT_PRIORITIES = [
  { value: 'normal',    label: 'Normal' },
  { value: 'important', label: 'Important' },
  { value: 'critical',  label: 'Critical' },
];

export const ANNOUNCEMENT_SCOPES = [
  { value: 'company',    label: 'Company-wide' },
  { value: 'department', label: 'Department' },
  { value: 'role',       label: 'Role' },
  { value: 'individual', label: 'Individual' },
];

File: Backend/app/routes/announcements.py
"""Announcements API routes.

All business logic lives in app.services.announcement_service.
This file is intentionally thin: auth, input binding, service call, serialise.
"""
from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.core.deps import get_current_user
from app.db.session import get_db
from app.models.employee import Employee
from app.schemas.announcement import (
    AnnouncementAnalyticsOut,
    AnnouncementCreateIn,
    AnnouncementOut,
    AnnouncementReadOut,
    AnnouncementUpdateIn,
)
from app.services import announcement_service as svc

router = APIRouter(prefix="/announcements", tags=["announcements"])


# ---------------------------------------------------------------------------
# Feed endpoints (employee / manager)
# ---------------------------------------------------------------------------

@router.get("", summary="List announcements visible to the current user")
def list_announcements(
    mode:        str  = Query("feed", description="'feed' for personal view, 'manage' for admin table"),
    unread_only: bool = Query(False),
    category:    Optional[str] = Query(None),
    priority:    Optional[str] = Query(None),
    status:      Optional[str] = Query(None, description="admin mode only: draft|published|archived"),
    search:      Optional[str] = Query(None),
    limit:       int  = Query(20, ge=1, le=100),
    offset:      int  = Query(0, ge=0),
    db: Session = Depends(get_db),
    user: Employee = Depends(get_current_user),
):
    role = user.role.name.lower() if user.role else "employee"

    if mode == "manage" and role in ("admin", "hr"):
        items, total = svc.list_for_admin(
            db, user,
            status_filter=status,
            category=category,
            priority=priority,
            search=search,
            limit=limit,
            offset=offset,
        )
    elif role == "manager":
        items, total = svc.list_for_manager(
            db, user,
            unread_only=unread_only,
            category=category,
            search=search,
            limit=limit,
            offset=offset,
        )
    else:
        items, total = svc.list_for_employee(
            db, user,
            unread_only=unread_only,
            category=category,
            priority=priority,
            search=search,
            limit=limit,
            offset=offset,
        )

    return {"total": total, "items": items}


@router.get("/latest", summary="Latest announcements visible to the current user (lightweight)")
def latest_announcements(
    limit: int = Query(5, ge=1, le=25),
    db: Session = Depends(get_db),
    user: Employee = Depends(get_current_user),
):
    """Convenience endpoint for dashboard widgets (e.g. the Admin "System
    Narrative" card). Returns a bare list sorted by pin-first then most
    recent — same audience filtering as the main feed, just a thinner
    payload and no pagination envelope.
    """
    role = user.role.name.lower() if user.role else "employee"
    if role in ("admin", "hr"):
        items, _ = svc.list_for_admin(
            db, user,
            status_filter="published",
            limit=limit, offset=0,
        )
    elif role == "manager":
        items, _ = svc.list_for_manager(
            db, user,
            unread_only=False, limit=limit, offset=0,
        )
    else:
        items, _ = svc.list_for_employee(
            db, user,
            unread_only=False, limit=limit, offset=0,
        )
    return items


@router.get("/unread-count", summary="Count unread announcements for current user")
def unread_count(
    db: Session = Depends(get_db),
    user: Employee = Depends(get_current_user),
):
    return {"count": svc.unread_count_for(db, user)}


@router.get("/{ann_id}", response_model=AnnouncementOut, summary="Get a single announcement")
def get_announcement(
    ann_id: int,
    db: Session = Depends(get_db),
    user: Employee = Depends(get_current_user),
):
    return svc.get_single(db, ann_id, user)


# ---------------------------------------------------------------------------
# Create / Update / Delete
# ---------------------------------------------------------------------------

@router.post("", response_model=AnnouncementOut, status_code=201, summary="Create a new announcement")
def create_announcement(
    payload: AnnouncementCreateIn,
    db: Session = Depends(get_db),
    user: Employee = Depends(get_current_user),
):
    ann = svc.create_announcement(db, user, payload)
    return svc.get_single(db, ann.id, user)


@router.put("/{ann_id}", response_model=AnnouncementOut, summary="Update an announcement")
def update_announcement(
    ann_id: int,
    payload: AnnouncementUpdateIn,
    db: Session = Depends(get_db),
    user: Employee = Depends(get_current_user),
):
    ann = svc.update_announcement(db, ann_id, user, payload)
    return svc.get_single(db, ann.id, user)


@router.delete("/{ann_id}", status_code=204, summary="Delete an announcement")
def delete_announcement(
    ann_id: int,
    db: Session = Depends(get_db),
    user: Employee = Depends(get_current_user),
):
    svc.delete_announcement(db, ann_id, user)


# ---------------------------------------------------------------------------
# Lifecycle transitions
# ---------------------------------------------------------------------------

@router.post("/{ann_id}/publish", response_model=AnnouncementOut, summary="Publish an announcement")
def publish_announcement(
    ann_id: int,
    db: Session = Depends(get_db),
    user: Employee = Depends(get_current_user),
):
    ann = svc.publish_announcement(db, ann_id, user)
    return svc.get_single(db, ann.id, user)


@router.post("/{ann_id}/archive", response_model=AnnouncementOut, summary="Archive an announcement")
def archive_announcement(
    ann_id: int,
    db: Session = Depends(get_db),
    user: Employee = Depends(get_current_user),
):
    ann = svc.archive_announcement(db, ann_id, user)
    return svc.get_single(db, ann.id, user)


# ---------------------------------------------------------------------------
# Read / Acknowledgement tracking
# ---------------------------------------------------------------------------

@router.post("/{ann_id}/read", response_model=AnnouncementReadOut, summary="Mark announcement as read")
def mark_read(
    ann_id: int,
    db: Session = Depends(get_db),
    user: Employee = Depends(get_current_user),
):
    return svc.mark_read(db, ann_id, user)


@router.post("/{ann_id}/acknowledge", response_model=AnnouncementReadOut, summary="Acknowledge an announcement")
def acknowledge(
    ann_id: int,
    db: Session = Depends(get_db),
    user: Employee = Depends(get_current_user),
):
    return svc.acknowledge(db, ann_id, user)


# ---------------------------------------------------------------------------
# Analytics
# ---------------------------------------------------------------------------

@router.get("/{ann_id}/analytics", response_model=AnnouncementAnalyticsOut, summary="Read/ack stats")
def analytics(
    ann_id: int,
    db: Session = Depends(get_db),
    user: Employee = Depends(get_current_user),
):
    return svc.get_analytics(db, ann_id, user)
